src/utils.c

```diff
@@ -128,6 +128,16 @@ char *get_album_cover_file(const char *location, const char *format) {
 	if (location == NULL)
 		return NULL;
 
+	/* show cover cue image file */
+	char *location_t;
+	location_t = strdup(location);
```

Comment:
Instead of creating another string duplicate, I would do something like this:
```c
	bool is_cue_file = false;
	if (strncmp(location, "cue://", 6) == 0) {
		is_cue_file = true;
		location += 6;
	}
```

Before diropen:
```c
	if (is_cue_file)
		strcat(fname, "/..");
```

Reply:
if you think about it, you can remove everything superfluous and leave only:
```c
if (strncmp(location, "cue://", 6) == 0) {
		location += 6;
                location = dirname(location);
}
```

Reply:
No, you cannot, because `dirname()` modifies it's argument and the location is `const`.

Reply:
And please add some comment why this extra `dirname()` is required.

Reply:
> No, you cannot, because `dirname()` modifies it's argument and the location is `const`.

Yes, when playing image files, this is good, as it fixes the path. Cmus when playing `*.cue` file images passes cmusfm this path:
```
cue:///home/user/music/Abba/Abba.cue/1
```
here `cue://` is a sign of playing a cue image, `...../Abba.cue/1` **1** at the end of the path, this is the track number in the image of the `*.cue`  file.
Function `dirname()` in this case returns the correct path, cutting off the track number (in this example it is 1).
```
cue:///home/user/music/Abba/Abba.cue
```
Here `dirname()` is needed to remove the `1` at the end of the path, the next call to `dirname()` returns the path:
```
.../Abba/
```
cutting off `Abba.cue`  this will be the album cover search path.

Modifying `dirname()` `const` does not matter here, since the path was initially incorrect. Maybe it's bad programming style, I don't know.
Сan be written correctly like this, and we get the same result.
```c
if (is_cue_file)
		strcat(fname, "/../..");
```

Reply:
> cutting off Abba.cue this will be the album cover search path.

OK, now I can see the whole picture :) So, indeed second `dirname()` is required. It's not possible to add "/.." because the Abba.cue is not a directory but a file.

> Modifying dirname() const does not matter here, since the path was initially incorrect. Maybe it's bad programming style, I don't know.

Yes, you can not use dirname on location variable because it's defined as const. It's a bad programming style which in some cases might lead to program crash.

Reply:
I want to thank you for giving me your time and for doing my training :)

then your code will work:
```c
bool is_cue_file = false;
	if (strncmp(location, "cue://", 6) == 0) {
		is_cue_file = true;
		location += 6;
	}
```
instead:

```c
	if (is_cue_file)
		strcat(fname, "/..");
```
use `if (is_cue_file)` here:
```c 
        tmp = strdup(location);                                                                                                                                                       
        if (is_cue_file)                                                                                                                                                    
                strcpy(tmp, dirname(tmp));                                                                                                                                            
        strcpy(fname, dirname(tmp));                                                                                                                                                                                                                                                                                            
        free(tmp);     
```

Reply:
> strcpy(tmp, dirname(tmp));

This code might be problematic. At first glance it looks OK, but you have to understand how the `dirname()` might be implemented to know the implications. The `dirname()` might modify given argument in place (hence the argument is not const) and might return address of given parameter - it might return address of `tmp`. So, you will have `strcpy(tmp, tmp)`. In the documentation for `strcpy()` you will read: "The  strcpy()  function copies the string pointed to by src, including the terminating null byte ('\0'), to the buffer pointed to by dest.  The strings may not overlap, and the destination string dest must be large enough to receive the copy."

Reply:
OK, If I understand correctly, then then can just use `dirname(fname)`
```c
  	tmp = strdup(location);
	strcpy(fname, dirname(tmp));
		if (is_cue_file) {
                        dirname(fname);
	free(tmp);
```
checked the code, everything works

Reply:
Yes, I think that such approach would be OK.